### Binning (Discretization)

Binning is the process of converting continuous numerical data into categorical groups or bins. Instead of using exact values like 500, 800, 1000, 1200, 1500, we group them into categories.

**Example:**
- 0 - 600 sq ft → Small House
- 600 - 1200 sq ft → Medium House
- 1200+ sq ft → Large House

This is called discretization.


### Why We Need Binning?

1. **Reduce Noise** - Small numerical changes don't always matter.
   - Example: 1001 sq ft vs 1003 sq ft is almost same. Binning treats them equally.

2. **Cleaner Patterns** - Models sometimes learn better with ranges instead of exact values.

3. **Business Logic** - Makes it easier for non-technical people to understand.


### Types of Binning

**1. Manual Binning (Domain-Based)**
You decide the ranges based on your knowledge or business rules.

**2. Equal-Width Binning**
Divides data into bins of equal size.

**3. Equal-Frequency Binning**
Divides data into bins with equal number of observations.


### Manual Binning Example

Let's create bins for the 'area' feature:

| Area Range | Category |
|------------|----------|
| 0 - 3000 | Very Small |
| 3000 - 6000 | Small |
| 6000 - 9000 | Medium |
| 9000 - 12000 | Large |
| 12000+ | Very Large |



## Step by Step Impletation.


# Step 1: Load cleaned house data.

In [1]:
import pandas as pd
df=pd.read_csv("../Data/Housing_cleaned.csv")
df.head()

,price,area,bedrooms,bathrooms,stories,mainroad,parking,furnishingstatus
0,9100000,6000,4,1,2,yes,2,semi-furnished
1,9100000,6600,4,2,2,yes,1,unfurnished
2,8890000,4600,3,2,2,yes,2,furnished
3,8855000,6420,3,2,2,yes,1,semi-furnished
4,8750000,4320,3,1,2,yes,2,semi-furnished


# Step 2: Basic analysis.

In [3]:
df.info()
df.describe()

<class 'pandas.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   price             463 non-null    int64
 1   area              463 non-null    int64
 2   bedrooms          463 non-null    int64
 3   bathrooms         463 non-null    int64
 4   stories           463 non-null    int64
 5   mainroad          463 non-null    str  
 6   parking           463 non-null    int64
 7   furnishingstatus  463 non-null    str  
dtypes: int64(6), str(2)
memory usage: 35.7 KB


,price,area,bedrooms,bathrooms,stories,parking
count,4.630000e+02,463.000000,463.000000,463.000000,463.000000,463.000000
mean,4.373745e+06,4766.583153,2.857451,1.220302,1.615551,0.587473
std,1.465136e+06,1795.315000,0.658924,0.440211,0.626901,0.776003
min,1.750000e+06,1650.000000,1.000000,1.000000,1.000000,0.000000
25%,3.342500e+06,3490.000000,2.000000,1.000000,1.000000,0.000000
50%,4.200000e+06,4300.000000,3.000000,1.000000,2.000000,0.000000
75%,5.222000e+06,6000.000000,3.000000,1.000000,2.000000,1.000000
max,9.100000e+06,10500.000000,4.000000,3.000000,3.000000,2.000000


# Step 3: Binning on Area

In [4]:
# Define bins
bins=[0,2500,4000,6000,8000,10000]
labels=["very_small","small","medium","large","very_large"]
df["area_bin"]=pd.cut(df["area"],bins=bins,labels=labels)
df.head(10)

,price,area,bedrooms,bathrooms,stories,mainroad,parking,furnishingstatus,area_bin
0,9100000,6000,4,1,2,yes,2,semi-furnished,medium
1,9100000,6600,4,2,2,yes,1,unfurnished,large
2,8890000,4600,3,2,2,yes,2,furnished,medium
3,8855000,6420,3,2,2,yes,1,semi-furnished,large
4,8750000,4320,3,1,2,yes,2,semi-furnished,medium
5,8680000,7155,3,2,1,yes,2,unfurnished,large
6,8645000,8050,3,1,1,yes,1,furnished,very_large
7,8645000,4560,3,2,2,yes,1,furnished,medium
8,8575000,8800,3,2,2,yes,2,furnished,very_large
9,8540000,6540,4,2,2,yes,2,furnished,large


# Step 4: Convert price to Lakhs



In [5]:
df["price"]=df["price"]/100000
df.head()

,price,area,bedrooms,bathrooms,stories,mainroad,parking,furnishingstatus,area_bin
0,91.00,6000,4,1,2,yes,2,semi-furnished,medium
1,91.00,6600,4,2,2,yes,1,unfurnished,large
2,88.90,4600,3,2,2,yes,2,furnished,medium
3,88.55,6420,3,2,2,yes,1,semi-furnished,large
4,87.50,4320,3,1,2,yes,2,semi-furnished,medium


# Step 5: Binning on Price

In [6]:
bins=[0, 25,40,60,80,100]
labels=["low-Budget","medium-budget","high-budget","luxury-budget","ultra-luxury-budget"]
df["price_bin"]=pd.cut(df["price"],bins=bins,labels=labels)
df.head(10)

,price,area,bedrooms,bathrooms,stories,mainroad,parking,furnishingstatus,area_bin,price_bin
0,91.00,6000,4,1,2,yes,2,semi-furnished,medium,ultra-luxury-budget
1,91.00,6600,4,2,2,yes,1,unfurnished,large,ultra-luxury-budget
2,88.90,4600,3,2,2,yes,2,furnished,medium,ultra-luxury-budget
3,88.55,6420,3,2,2,yes,1,semi-furnished,large,ultra-luxury-budget
4,87.50,4320,3,1,2,yes,2,semi-furnished,medium,ultra-luxury-budget
5,86.80,7155,3,2,1,yes,2,unfurnished,large,ultra-luxury-budget
6,86.45,8050,3,1,1,yes,1,furnished,very_large,ultra-luxury-budget
7,86.45,4560,3,2,2,yes,1,furnished,medium,ultra-luxury-budget
8,85.75,8800,3,2,2,yes,2,furnished,very_large,ultra-luxury-budget
9,85.40,6540,4,2,2,yes,2,furnished,large,ultra-luxury-budget


# Step 6: Convert Bins into Feature vector.(One-Hot-Encoding)



### Encoding

Encoding is the process of converting categorical values into numerical values so that machine learning models can train on them.


### One-Hot Encoding

Each category becomes a separate binary feature (0 or 1). Only one feature is active at a time.

**Example:**

Houses = ["Small", "Medium", "Large"]

**One-Hot Encoding:**

| House   | Small | Medium | Large |
|---------|-------|--------|-------|
| Small   | 1     | 0      | 0     |
| Medium  | 0     | 1      | 0     |
| Large   | 0     | 0      | 1     |

### Why Use One-Hot Encoding?

1. **No Order Assumption** - Does not imply any ranking between categories
2. **Works with All Models** - Linear, Tree-based, Neural Networks
3. **Simple to Implement** - Easy to understand and use
4. **Preserves Information** - No loss of categorical information


# Step 6: Encode on Area bin and Price Bin.

In [7]:
df["mainroad"]=df["mainroad"].map({"yes":1,"no":0})
df_encoded=pd.get_dummies(df,columns=["area_bin","price_bin","furnishingstatus"],drop_first=True)
df_encoded.head()

,price,area,bedrooms,bathrooms,stories,mainroad,parking,area_bin_small,area_bin_medium,area_bin_large,area_bin_very_large,price_bin_medium-budget,price_bin_high-budget,price_bin_luxury-budget,price_bin_ultra-luxury-budget,furnishingstatus_semi-furnished,furnishingstatus_unfurnished
0,91.00,6000,4,1,2,1,2,False,True,False,False,False,False,False,True,True,False
1,91.00,6600,4,2,2,1,1,False,False,True,False,False,False,False,True,False,True
2,88.90,4600,3,2,2,1,2,False,True,False,False,False,False,False,True,False,False
3,88.55,6420,3,2,2,1,1,False,False,True,False,False,False,False,True,True,False
4,87.50,4320,3,1,2,1,2,False,True,False,False,False,False,False,True,True,False


# Step 7: Convert True/False to 1/0.

In [8]:
df_encoded = df_encoded.replace({True: 1, False: 0})
df_encoded.head()

,price,area,bedrooms,bathrooms,stories,mainroad,parking,area_bin_small,area_bin_medium,area_bin_large,area_bin_very_large,price_bin_medium-budget,price_bin_high-budget,price_bin_luxury-budget,price_bin_ultra-luxury-budget,furnishingstatus_semi-furnished,furnishingstatus_unfurnished
0,91.00,6000,4,1,2,1,2,0,1,0,0,0,0,0,1,1,0
1,91.00,6600,4,2,2,1,1,0,0,1,0,0,0,0,1,0,1
2,88.90,4600,3,2,2,1,2,0,1,0,0,0,0,0,1,0,0
3,88.55,6420,3,2,2,1,1,0,0,1,0,0,0,0,1,1,0
4,87.50,4320,3,1,2,1,2,0,1,0,0,0,0,0,1,1,0


## Feature Engineering - Part II Complete

### Summary of Part II

In this notebook, we completed the following feature engineering techniques:

| Step | Technique | Description |
|------|-----------|-------------|
| 1 | Data Loading | Loaded cleaned housing dataset (463 records) |
| 2 | Basic Analysis | Checked data types and statistics |
| 3 | Binning (Area) | Converted area into 5 categories: very_small, small, medium, large, very_large |
| 4 | Price Conversion | Converted price to lakhs (divided by 100,000) |
| 5 | Binning (Price) | Converted price into 5 budget categories |
| 6 | One-Hot Encoding | Converted categorical bins into binary features |
| 7 | Boolean to Integer | Converted True/False to 1/0 |

### Saving Encoded Data and  Final Dataset Shape


In [10]:
df_encoded.to_csv("../Data/Housing_cleaned_encoded.csv",index=False)
print("DataFrame has been encoded and saved to Housing_cleaned_encoded.csv")

DataFrame has been encoded and saved to Housing_cleaned_encoded.csv


In [11]:
print(f"Final dataset shape: {df_encoded.shape}")
print(f"Features: {df_encoded.shape[1]}")
print(f"Records: {df_encoded.shape[0]}")

Final dataset shape: (463, 17)
Features: 17
Records: 463
